homeassistant/components/unifiprotect/data.py

```diff
@@ -0,0 +1,148 @@
+
+        self.last_update_success = False
+        self.access_tokens: dict[str, collections.deque] = {}
+        self.api = protect
+
+    @property
+    def disable_stream(self) -> bool:
+        """Check if RTSP is disabled."""
+        return self._entry.options.get(CONF_DISABLE_RTSP, False)
+
+    async def async_setup(self) -> None:
+        """Subscribe and do the refresh."""
+        self._unsub_websocket = self.api.subscribe_websocket(
+            self._async_process_ws_message
+        )
+        await self.async_refresh()
+
+    async def async_stop(self, *args: Any) -> None:
+        """Stop processing data."""
+        if self._unsub_websocket:
+            self._unsub_websocket()
+            self._unsub_websocket = None
+        if self._unsub_interval:
+            self._unsub_interval()
+            self._unsub_interval = None
+        await self.api.async_disconnect_ws()
+
+    async def async_refresh(self, *_: Any, force: bool = False) -> None:
+        """Update the data."""
+
+        # if last update was failure, force until success
+        if not self.last_update_success:
+            force = True
+
+        try:
+            updates = await self.api.update(force=force)
+        except NvrError:
+            if self.last_update_success:
+                _LOGGER.exception("Error while updating")
+            self.last_update_success = False
+            # manually trigger update to mark entities unavailable
+            self._async_process_updates(self.api.bootstrap)
+        except NotAuthorized:
+            await self.async_stop()
+            _LOGGER.exception("Reauthentication required")
+            self._entry.async_start_reauth(self._hass)
+            self.last_update_success = False
+        else:
+            self.last_update_success = True
+            self._async_process_updates(updates)
+
+    @callback
+    def _async_process_ws_message(self, message: WSSubscriptionMessage) -> None:
+        if message.new_obj.model in DEVICES_WITH_ENTITIES:
+            self.async_signal_device_id_update(message.new_obj.id)
+
+    @callback
+    def _async_process_updates(self, updates: Bootstrap | None) -> None:
+        """Process update from the protect data."""
+
+        # Websocket connected, use data from it
+        if updates is None:
+            return
+
+        self.async_signal_device_id_update(self.api.bootstrap.nvr.id)
+        for device_type in DEVICES_THAT_ADOPT:
+            attr = f"{device_type.value}s"
+            devices: dict[str, ProtectDeviceModel] = getattr(self.api.bootstrap, attr)
+            for device_id in devices.keys():
+                self.async_signal_device_id_update(device_id)
+
+    @callback
+    def async_subscribe_device_id(
+        self, device_id: str, update_callback: CALLBACK_TYPE
+    ) -> CALLBACK_TYPE:
+        """Add an callback subscriber."""
+        if not self._subscriptions:
+            self._unsub_interval = async_track_time_interval(
```

Comment:
We may be able to leverage our data update coordinator helper. It supports manual updates too. See eg the Shelly integration.

Reply:
Martin is referencing this:

https://developers.home-assistant.io/docs/integration_fetching_data/#coordinated-single-api-poll-for-data-for-all-entities

Reply:
We can't use the coordinator in this case because it doesn't implement selective callbacks.  We don't want to write state on 300+ entities (some people even have larger installs) every time a motion sensor updates.

Reply:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/data.py#L142 We need the per device id callbacks since the websocket can be quite active.

Reply:
The "polling" aspect of this integration is mostly just to ensure the WS is still active. Polling updates from the API is a last result if the Websocket cannot stay active.